### ISSUE TITLE
Kiali 1.36 upgrade from 1.32 service selector incorrect

### ISSUE THREAD

**Issue by mjnagel** [kiali/kiali#4143]
**Issue Description**
We are noticing that when upgrading from 1.32 to 1.36 the service selectors seem to stick around:

```yaml
    selector:
      app: kiali
      app.kubernetes.io/instance: kiali
      app.kubernetes.io/name: kiali
      version: v1.32.0
```

This causes issues with hitting the service obviously. We are installing from the helm chart, so its possible this is related to the helm chart rather than the operator?

**Versions used**
Kiali: 1.32.0 -> 1.36.0
Istio: 1.8.4
Kubernetes flavour and version: k3s version v1.20.4-k3s1

**To Reproduce**
1. Helm install the 1.32.0 operator + instance
2. Helm upgrade to the 1.36.0 operator + instance
3. Check the kiali service that has been created
4. See the extra/wrong selectors.

**Expected behavior**
The service should be updated on upgrades.

**Comment by jmazzitelli** [comment]
So we can attempt to replicate this as exactly as possible, please provide the exact commands you are executing to both helm install and helm upgrade the operator.

**Comment by jmazzitelli** [comment]
Here's some docs that might help: https://kiali.io/documentation/latest/installation-guide/#_upgrading_kiali

And I will assume you are not deleting the Kiali CR during the upgrade. So you have a Kiali server installed, and then you are upgrading the operator. It might be more successful if you delete the existing Kiali CRs first before upgrading the operator. Otherwise, if you, for example, have set spec.deployment.image_version in the CR, the Kiali server will remain at that version you specified. There might be some other side-effects when keeping old CRs around during an operator upgrade.

**Comment by mjnagel** [comment]
@jmazzitelli originally I was doing this on a chart install with a lot of values changes. I was just able to reproduce with a simpler setup however and will provide the commands used below. Essentially we are upgrading both the "instance" and the operator at the same time. From port forwarding/validating the image versions we can see that everything comes up healthy and the images have upgrade but the service is incorrect.

With istio 1.8.5 installed in the istio-system namespace:

For initial install 1.32:
```
helm install --create-namespace --namespace kiali --set cr.spec.auth.strategy="anonymous" --repo https://kiali.org/helm-charts --version 1.32.0 --set cr.create=true --set cr.spec.istio_namespace=istio-system --set cr.spec.deployment.image_version=v1.32.0 kiali-operator kiali-operator
```

After completion of install/operator reconciling the Kiali CR the svc shows these selectors:
```
    selector:
      app: kiali
      version: v1.32.0
```

For upgrade to 1.36:
```
helm upgrade -i --create-namespace --namespace kiali --set cr.spec.auth.strategy="anonymous" --repo https://kiali.org/helm-charts --version 1.36.0 --set cr.create=true --set cr.spec.istio_namespace=istio-system kiali-operator kiali-operator
```

After completion of upgrade/operator reconciling the Kiali CR the svc shows these selectors:
```
    selector:
      app: kiali
      app.kubernetes.io/instance: kiali
      app.kubernetes.io/name: kiali
      version: v1.32.0
```

This results in errors accessing the service as that version tag is not on the active deployment. Any thoughts on what we're doing wrong here or anything we could do to work around this? It looks like the new selector tags get added/patched in with the old ones still remaining.

**Comment by jmazzitelli** [comment]
OK. I know what this is. The "version" field has been removed from the selector (happened in [this PR](https://github.com/kiali/kiali-operator/commit/b6f243cc70136657d7f19209312790f51ce9a221#diff-c44c399bcc43501fcc4d11c804f99ca309c269d3b8516289fd853b84bf0cc1e5L34) which went into [Kiali v1.35.0](https://github.com/kiali/kiali-operator/blob/v1.35.0/roles/default/kiali-deploy/templates/kubernetes/service.yaml#L27-L29)). So this problem only happens if you upgrade from a pre-v1.35 operator to a v1.35-or-higher operator. Once you get past v1.35, this won't happen.

It looks like Kubernetes patch is not removing that now-defunct "version" field when applied to an existing service that already had it.

There are two workarounds you can do for this.

1. Delete the Kiali CR prior to the upgrade - this uninstalls the Kiali Server. The upgraded operator will create a new Kiali deployment.

or

2. Manually patch the service after the upgrade is complete and the service has been updated:
```
kubectl patch service kiali -n kiali --type=json -p='[{"op":"remove","path":"/spec/selector/version"}]'
```

**Comment by mjnagel** [comment]
Perfect that helps a lot, I wondered if it was something like that with some manual pieces required for the upgrade - just was hoping that the operator handled it automagically.

We can definitely do one of those two to get around it, likely the patch option will be best for us. And unless I'm mistaken we should also remove the `app` selector from the service since that was removed in 1.35 as well?

**Comment by jmazzitelli** [comment]
> And unless I'm mistaken we should also remove the `app` selector from the service since that was removed in 1.35 as well?

You could do that - it wouldn't hurt. But the existence of "app" in the selector won't cause a problem because the operator [still applies the "app" label to the Kiali resources, including the deployment/pod](https://github.com/kiali/kiali-operator/blob/5b07214db85f6f892cf8937239416694873e476e/roles/default/kiali-deploy/tasks/main.yml#L496-L499) - so that selector will still match. But you are right - "app" has been removed from the selector in the upgraded service, so for completeness, you can remove that, too.

**Comment by mjnagel** [comment]
Yep, tracking! I think we'll remove it for completeness/to have it match what a "clean install" would look like. Thanks for the assistance on this 😄

**Comment by jmazzitelli** [comment]
I was going to fix this - it looked "easy" - see draft PR: https://github.com/kiali/kiali-operator/pull/348

But! Then I hit what appears to be a [bug in ansible](https://github.com/ansible-collections/kubernetes.core/issues/161), so, it isn't easy :)

For now, we'll have to just recommend those workarounds mentioned above. If there is a workaround to that ansible bug, we can revisit the fix.

**Comment by mjnagel** [comment]
Ah fun, no worries. We plan to tackle this with a helm post-upgrade job that runs the patch commands so that it can all be "automated" with the helm upgrade. Not too bad to do it this way, although it would definitely be nice to get the operator to handle this.

If you want I can reopen this issue for future working since it sounds like it was unintended and more of a bug than an intended upgrade path?

**Comment by jmazzitelli** [comment]
We can leave this open. If we can get a fix to work, we'll do that.

**Comment by stevehipwell** [comment]
This is a pretty major issue, the whole point of an operator is to handle this "operations" stuff. Having to manually patch a deployment after the operator runs is not OK. I'm not sure what the actual issue with fixing this is as it used to work fine?

**Comment by jmazzitelli** [comment]
> I'm not sure what the actual issue with fixing this

Did you see my comment where I describe and link to the actual issues? https://github.com/kiali/kiali/issues/4143#issuecomment-873170499

**Comment by stevehipwell** [comment]
@jmazzitelli I saw the issue and am surprised that there isn't a custom action hook in an Ansible operator? If not and until the issue is resolved what is the problem with keeping the version up to date so we don't get a broken service each time the operator runs?

**Comment by jmazzitelli** [comment]
I want to more formally declare what the problem is.

The problem only shows up if:

(a) You have already installed Kiali server using operator v1.34 or earlier, and...
(b) You have not deleted the Kiali CR (i.e. you have not uninstalled that Kiali server), and...
(c) You upgrade the operator (and hence the existing Kiali server) to v1.35 or higher.

If you delete the Kiali CRs prior to upgrading the operator, this isn't a problem.

If you are starting from v1.35 or higher (and, say, upgrade to v1.36.1) or if you are starting before v1.35 and are upgrading to before v1.35, this also isn't a problem.

So, in short, upgrading an existing Kiali server from pre-v1.35 to 1.35+ is a problem.

**Comment by jmazzitelli** [comment]
I think I have a [solution](https://github.com/kiali/kiali-operator/pull/348) for this - its not the best but, it does seem to work.
